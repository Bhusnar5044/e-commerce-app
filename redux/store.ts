import { configureStore } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { auth, initialState as authInitialState } from './features/authAlice';
import { cart, initialState as cartInitialState } from './features/cartSlice';
import { modal, initialState as modalInitialState } from './features/modalSlice';

export const initialState = {
    auth: authInitialState,
    cart: cartInitialState,
    modal: modalInitialState,
};

export const store = configureStore({
    reducer: { auth: auth.reducer, cart: cart.reducer, modal: modal.reducer },
});

export type AppState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<AppState> = useSelector;
