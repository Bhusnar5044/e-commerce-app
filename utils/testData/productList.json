[
    {
        "id": "1",
        "category": { "id": "1", "name": "Glasses" },
        "price": "12",
        "isFeatured": true,
        "name": "Dummy",
        "size": { "id": "1", "name": "S", "value": "12" },
        "color": { "id": "1", "name": "Gray", "value": "gray" },
        "images": [{ "id": "1", "url": "" }]
    }
]
