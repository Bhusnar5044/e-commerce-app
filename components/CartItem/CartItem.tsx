import { X } from 'lucide-react';
import Image from 'next/image';

import Currency from '@/components/ui/Currency';
import IconButton from '@/components/ui/IconButton';
import { useAppDispatch } from '@/redux';
import { removeItem } from '@/redux/features/cartSlice';
import { Product } from '@/types';
import { FC, memo } from 'react';

interface CartItemProps {
    data: Product;
}

export const CartItem: FC<CartItemProps> = memo(({ data }) => {
    const dispatch = useAppDispatch();

    const onRemove = () => {
        dispatch(removeItem(data.id));
    };

    return (
        <li className="flex py-6 border-b">
            <div className="relative h-24 w-24 rounded-md overflow-hidden sm:h-48 sm:w-48">
                <Image fill src={data.thumbnail} alt="" className="object-cover object-center" />
            </div>
            <div className="relative ml-4 flex flex-1 flex-col justify-between sm:ml-6">
                <div className="absolute z-10 right-0 top-0">
                    <IconButton onClick={onRemove} icon={<X size={15} />} />
                </div>
                <div className="relative pr-9 sm:grid sm:grid-cols-2 sm:gap-x-6 sm:pr-0">
                    <div className="flex justify-between">
                        <p className=" text-lg font-semibold text-black">{data.title}</p>
                    </div>
                    <Currency value={data.price} />
                </div>
            </div>
        </li>
    );
});

CartItem.displayName = 'CartItem';
