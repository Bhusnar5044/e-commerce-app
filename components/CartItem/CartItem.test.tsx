import { initialState } from '@/redux';
import { fireEvent, mockStore, render, renderWithStore, screen } from '@test-utils';
import * as Products from '@utils/testData/productList.json';
import { Provider } from 'react-redux';
import { CartItem } from './CartItem';

const store = mockStore({ ...initialState });
describe('CartItem', () => {
    it('should render properly', () => {
        const { container } = renderWithStore(<CartItem data={Products[0]} />);
        expect(container).toMatchSnapshot();
        expect(screen.getByText(Products[0].title)).toBeInTheDocument();
    });

    it('should dispatch remove cart item action on click of close icon', () => {
        render(
            <Provider store={store}>
                <CartItem data={Products[0]} />
            </Provider>,
        );
        const closeBtn = screen.getByTitle(/iconButton/);
        fireEvent.click(closeBtn);
        expect(store.getActions()).toEqual([{ payload: Products[0].id, type: 'cart/removeItem' }]);
    });
});
