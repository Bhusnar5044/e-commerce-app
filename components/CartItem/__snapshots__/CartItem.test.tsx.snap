// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem should render properly 1`] = `
<div>
  <li
    class="flex py-6 border-b"
  >
    <div
      class="relative h-24 w-24 rounded-md overflow-hidden sm:h-48 sm:w-48"
    >
      <img
        alt=""
        class="object-cover object-center"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=3840&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F1%2Fthumbnail.jpg&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <div
      class="relative ml-4 flex flex-1 flex-col justify-between sm:ml-6"
    >
      <div
        class="absolute z-10 right-0 top-0"
      >
        <button
          class="rounded-full flex items-center justify-center bg-white border shadow-md p-2 hover:scale-110 transition"
          title="iconButton"
        >
          <svg
            class="lucide lucide-x"
            fill="none"
            height="15"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6 6 18"
            />
            <path
              d="m6 6 12 12"
            />
          </svg>
        </button>
      </div>
      <div
        class="relative pr-9 sm:grid sm:grid-cols-2 sm:gap-x-6 sm:pr-0"
      >
        <div
          class="flex justify-between"
        >
          <p
            class=" text-lg font-semibold text-black"
          >
            iPhone 9
          </p>
        </div>
        <div
          class="font-semibold"
        >
          $549.00
        </div>
      </div>
    </div>
  </li>
</div>
`;
